### PR TITLE
fix: disable withdraw when balance is zero

### PR DESCRIPTION
**Pull request by tansawit**
## Overview
This fixes the withdraw flow so a user with zero balance cannot fetch a route and reach an enabled withdraw confirm state.

## Root Cause
The regression traces back to PR #203, which intended to avoid flashing `Insufficient balance` while balances were still loading. That intent was reasonable, but the implementation used `balance === undefined` to mean “still loading.” In this flow, `undefined` could also mean the balances had already loaded and the selected denom was simply absent because the user had zero balance. That let zero-balance withdraws bypass validation.

## Fix
After the first balance snapshot is available, a missing selected denom is now treated as `0` instead of “unknown.” The withdraw form uses that resolved balance to block invalid amounts and prevent route fetching, and the footer also verifies source-asset sufficiency before enabling confirm.

Fixes: [EXP-299](https://linear.app/initia-labs/issue/EXP-299/disable-withdraw-when-balance-is-0)

---

> [!NOTE]
> **Medium Risk**
> Changes transfer balance resolution and footer/route gating logic, which can affect whether users can proceed with deposit/withdraw flows and how errors/warnings display.
> 
> **Overview**
> Fixes a withdraw regression where a missing denom balance (`undefined`) could be treated as “still loading” and allow route fetching/confirmation with an effective zero balance.
> 
> Introduces `getResolvedTransferBalance` and `hasSufficientTransferBalance` to consistently treat missing balances as `0` after the first snapshot, updates `TransferFields` to gate validation/route queries/MAX behavior on the resolved balance, and centralizes footer error/warning selection via new `getTransferFooterStatus` (with added unit tests).
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 373b9e4c3da7e76987337d5692c98972e2cb0bda. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

## Summary by CodeRabbit

* **Bug Fixes**
  * Balance display now waits for the first balances snapshot before showing/validating amounts; MAX and insufficient-balance gating behave correctly.
  * Footer now more reliably surfaces insufficient-balance errors instead of being suppressed in fee/source-denom edge cases.

* **New Features**
  * Centralized footer status logic to consistently show errors or warnings about fees and source balance.

* **Tests**
  * Added tests covering balance-resolution and sufficiency checks and footer status outcomes.

### CODE REVIEW

File: packages/interwovenkit-react/src/pages/deposit/transferBalanceGate.ts [hunk 1]
```diff
@@ -15,3 +17,44 @@ export function getTransferBalanceBlocker({
...
+interface HasSufficientTransferBalanceArgs {
+  balance?: string
+  requiredAmount: string
+}
+
+function getNonNegativeAmount(value: string): BigNumber | null {
+  if (!value) return null
+
+  const amount = BigNumber(value)
+  if (!amount.isFinite() || amount.isNaN() || amount.lt(0)) return null
```
Comment (simcheolhwan) [inline packages/interwovenkit-react/src/pages/deposit/transferBalanceGate.ts]: `BigNumber.isFinite()` already returns `false` for `NaN`, so `isNaN()` is redundant here.

```suggestion
  if (!amount.isFinite() || amount.lt(0)) return null
```

File: packages/interwovenkit-react/src/pages/deposit/transferBalanceGate.test.ts [hunk 2]
```diff
@@ -21,3 +25,85 @@ describe("getTransferBalanceBlocker", () => {
...
+      }),
+    ).toBe(false)
+    expect(
+      hasSufficientTransferBalance({
+        balance: "1",
+        requiredAmount: "-1",
+      }),
+    ).toBe(false)
+  })
+})
```
Comment (simcheolhwan) [inline packages/interwovenkit-react/src/pages/deposit/transferBalanceGate.test.ts:118]: Missing a test for exact equality (`balance === requiredAmount`). This is the MAX button path -- if `gte` is accidentally changed to `gt`, MAX withdraw silently breaks.

```suggestion
  })

  it("returns true when balance exactly equals required amount", () => {
    expect(
      hasSufficientTransferBalance({
        balance: "100",
        requiredAmount: "100",
      }),
    ).toBe(true)
  })
})
```

File: packages/interwovenkit-react/src/pages/deposit/TransferFooter.tsx [hunk 3]
```diff
@@ -74,7 +75,6 @@ function selectFeeDenom({
 
   return feeOptions[0]?.amount[0].denom
 }
-
 interface FooterWithFeeProps extends Omit<Props, "gas" | "mode"> {
```
Comment (simcheolhwan) [inline packages/interwovenkit-react/src/pages/deposit/TransferFooter.tsx:80]: Blank line between `selectFeeDenom` and `FooterWithFeeProps` was accidentally removed.

```suggestion
}

interface FooterWithFeeProps extends Omit<Props, "gas" | "mode"> {
```

File: packages/interwovenkit-react/src/pages/deposit/TransferFooter.tsx [hunk 4]
```diff
@@ -170,7 +170,14 @@ const TransferFooterWithFee = ({
     sourceDenom: srcDenom,
     feeDetailsByDenom,
   })
-  const balanceError = feeDetails && !feeDetails.isSufficient ? "Insufficient balance" : undefined
+  const hasSourceBalance = hasSufficientTransferBalance({
+    balance: balancesByDenom.get(srcDenom) ?? "0",
+    requiredAmount: sourceSpendAmount,
+  })
```
Comment (simcheolhwan) [inline packages/interwovenkit-react/src/pages/deposit/TransferFooter.tsx:178]: `TransferFields` uses `getResolvedTransferBalance` to resolve a missing denom to `"0"`, but here the same semantic is expressed via inline `?? "0"`. Consider using the same function for consistency, or add a comment noting the suspense boundary guarantees balances are loaded.

```suggestion
  const hasSourceBalance = hasSufficientTransferBalance({
    balance: balancesByDenom.get(srcDenom) ?? "0", // suspense boundary ensures balances are loaded; missing denom means zero
    requiredAmount: sourceSpendAmount,
  })
```

File: packages/interwovenkit-react/src/pages/deposit/TransferFooter.tsx [hunk 5]
```diff
@@ -170,7 +170,14 @@ const TransferFooterWithFee = ({
...
+  const balanceError =
+    !hasSourceBalance || (feeDetails && !feeDetails.isSufficient)
+      ? "Insufficient balance"
+      : undefined
```
Comment (simcheolhwan) [inline packages/interwovenkit-react/src/pages/deposit/TransferFooter.tsx]: Pre-existing edge case: when `feeWarning` is set and `feeDenom === srcDenom`, `balanceError` is suppressed in favor of the warning. If the user has insufficient balance for the transfer itself but enough for the fee (e.g. balance=50, transfer=100, fee=10), they see the fee warning instead of "Insufficient balance".

Not blocking -- worth tracking as a follow-up since the new `!hasSourceBalance` check increases the surface area.

**Review by simcheolhwan — APPROVED** [review]
Looks good. Balance resolution and footer status extraction are clean.

One remaining suggestion: add tests for `getTransferFooterStatus` covering the different-denom and happy-path branches.
